queries/nix.scm

```diff
@@ -0,0 +1,146 @@
+;; Statements
+
+[
+    (binding)
+] @statement
+;; This matches assert statements as stand-alone statements. This
+;; is because it's common to have a series of assert statements in a file, which
+;; would otherwise all match as a single statement. See the playground/nix/asserts.nix
+;; file for an example.
+(assert_expression
+    "assert" @statement.start
+    condition: (_)
+    ";" @statement.end
+    body: (_)
```

Comment:
or this line. unless a `;` can appear later in this pattern?

Reply:
Right ya removing body is ok.

I'm still kind of back and forth what I want to do about `assert`, and I'm starting to lean towards not special casing these as statements and just leveraging `line` scope instead. But I guess I will explain it here a bit to see if you or @auscompgeek has any thoughts.

Basically the form of an assert is: `assert <condition>; <body>` where `body` is often another `assert` (see data/playground/asserts.nix). So if I don't explicitly add `;` as the `@statement.end`, if I say `take state` on the first `assert`, it will take all of the subsequent `asserts`. Maybe this is actually okay, but it felt weird at first.

Here is an example of what I originally thought I wanted, where I `take state` on the second `assert` in the chain of asserts and it just applies to the single line and basically ignores the `body` of the expression:

![image](https://github.com/cursorless-dev/cursorless/assets/13679876/09679dae-6ef2-4338-923e-a9f9c194e059)

On the contrary, this is what happens if I don't force the `assert` statement to end early by marking the `;` and having it keep the `body`, and say `take state` on the second `assert:

![image](https://github.com/cursorless-dev/cursorless/assets/13679876/8e2342b2-9168-4506-b197-954c54239553)

Also the caveat of special casing assert as a statement the way I do atm, is it will hijack a `take state` on a simple variable assignment, which is probably not wanted, for instance:

![image](https://github.com/cursorless-dev/cursorless/assets/13679876/da4fd14a-fc73-4dc0-9e17-11694ef9ff55)

Maybe that's okay, but also I suspect a bit unintuitive. I could special case to not have variable assignments with `assert` apply. But then say the variable assignment has chain of asserts like the one earlier in the function, I may suddenly want to access each assert as an individual statement :D 

I guess this is the type of thing that could always be tweaked in a future PR after getting some use out of it |or do you prefer not changing this type of behavior once they're in?

Reply:
fwiw for now I changed it so that assert doesn't match as a statement if part of a binding, so the `foo = assert` will not longer trigger.

Reply:
Ok a couple surprising things when exploring the `assert` flie:

## `"call"`

<img width="717" alt="image" src="https://github.com/cursorless-dev/cursorless/assets/755842/de884eda-e339-4e10-9ca2-2b753b749719">

To be honest, I wasn't expecting asserts to match "call" at all. It feels more like a statement to me than calling an "assert" function. But I could see that going either way cc/ @auscompgeek

The more surprising thing is that while the range seems reasonable (assuming we want asserts to be calls), the domain was quite surprising. It seems to contain everything after the assert statement. Hence the increasingly light color of each subsequent assert in the above screenshot.

I'm guessing that has something to do with this whole thing being pure functional, so it's kinda like how `let x in ...` technically includes the `...` in langs like clojure. But I'm not sure it's desirable behaviour for the "call" scope

This "call" in particular confused me:

<img width="437" alt="image" src="https://github.com/cursorless-dev/cursorless/assets/755842/8aa1b4ba-7ee1-4cf8-affe-fcf3a345ea3d">

Should it not include `"a"`? If not, should it really include the `;`? Is that assert statement even valid? I made it up 😅

## `"state"`

<img width="717" alt="image" src="https://github.com/cursorless-dev/cursorless/assets/755842/2ec34cf3-401d-46b1-8630-7baa077a5db9">

Agreed that `assert` should be a statement, and I'm happy with your ranges here. However, as with `"call"` above, it seems to be exhibiting the same surprising nested domain behaviour

And agreed that it shouldn't be a statement if it's the rhs of an assignment. Tbh I have no idea what that's supposed to do, but it doesn't look like a statement to me 😅

## `"branch"`

<img width="575" alt="image" src="https://github.com/cursorless-dev/cursorless/assets/755842/1ef1068d-71ab-463c-8fc6-8e54a9b65672">

This one was also surprising to me. I guess it makes more sense in a one-line assert? cc/ @auscompgeek 

## `"callee"`

<img width="584" alt="image" src="https://github.com/cursorless-dev/cursorless/assets/755842/eea50f9d-8b17-4ce6-808d-9feeaf14cac2">

As above, the fact this is a "callee", as well as the domain, were surprising to me

## `"condition"`

<img width="588" alt="image" src="https://github.com/cursorless-dev/cursorless/assets/755842/05440733-91d3-4603-8f45-948ec610c08f">

This one makes the most sense of any to me, even the domain. I guess if people see assert statements this way, then maybe the other domains make sense as well? Still not totally sure bout "call" tho

## Tests

Looks like there are no tests for assert

- [ ] Assert as statement
- [ ] Assert condition
- [ ] Assert as call, if we decide to keep that
- [ ] Assert callee, if we decide to keep that
- [ ] Assert branch, if we decide to keep that
- [ ] Negative test for assert not being statement when on rhs of assignment (say "cursorless record error" and then do "change state")